src/provider.rs

```diff
@@ -903,95 +893,135 @@ fn compute_outboard(
+    }
+
+    fn add_source(&mut self, source: DataSource) -> &mut Self {
+        self.sources.push(source);
+        self
+    }
 
-    Ok((db, entries, hash))
+    /// Create the actual collection.
+    pub async fn build(self) -> Result<(Database, Hash)> {
+        let (db, _, hash) = self.build_with_entries().await?;
+        Ok((db, hash))
+    }
+
+    /// Create the actual collection, including `ProvideResponseEntry`s.
+    /// Writing the data into the provided database.
+    async fn build_into_with_entries(
+        self,
+        db: &mut HashMap<Hash, BlobOrCollection>,
+    ) -> Result<(Vec<ProvideResponseEntry>, Hash)> {
+        // +1 is for the collection itself
+        let mut blobs = Vec::with_capacity(self.sources.len());
+        let mut total_blobs_size: u64 = 0;
+
+        // compute outboards in parallel, using tokio's blocking thread pool
+        let outboards = self.sources.into_iter().map(|data| {
+            let (path, name) = match data {
+                DataSource::File(path) => (path, None),
+                DataSource::NamedFile { path, name } => (path, Some(name)),
+            };
+            tokio::task::spawn_blocking(move || compute_outboard(path, name))
+        });
+        // wait for completion and collect results
+        let outboards = future::join_all(outboards)
+            .await
+            .into_iter()
+            .collect::<Result<Result<Vec<_>, _>, _>>()??;
+
+        // compute information about the collection
+        let entries = outboards
+            .iter()
+            .map(|(path, name, hash, outboard)| {
+                let name = name
+                    .as_ref()
+                    .cloned()
+                    .unwrap_or_else(|| path.display().to_string());
+                ProvideResponseEntry {
+                    name,
+                    hash: *hash,
+                    size: u64::from_le_bytes(outboard[..8].try_into().unwrap()),
+                }
+            })
+            .collect();
+
+        // insert outboards into the database and build collection
+        for (path, name, hash, outboard) in outboards {
+            debug_assert!(outboard.len() >= 8, "outboard must at least contain size");
+            let size = u64::from_le_bytes(outboard[..8].try_into().unwrap());
+            db.entry(hash).or_insert_with(|| {
+                BlobOrCollection::Blob(Data {
+                    outboard: Bytes::from(outboard),
+                    path: path.clone(),
+                    size,
+                })
+            });
+            total_blobs_size += size;
+            // if the given name is `None`, use the filename from the given path as the name
+            let name = name.unwrap_or_else(|| {
+                path.file_name()
+                    .and_then(|s| s.to_str())
+                    .unwrap_or_default()
+                    .to_string()
+            });
+            blobs.push(Blob { name, hash });
+        }
+
+        let c = Collection {
```

Comment:
I think it might be nice to canonicalize the collection here by calling sort on the blobs.

Not that we want to get overly religious regarding canonicalisation, but it helps a bit with tests, and if it is basically free why not do it.